kdeploy

```diff
@@ -0,0 +1,186 @@
+	    -u, --remote-user USER	User to be used for remote deploy via ssh/rsync
+	    -p, --remote-port PORT	Port to be used for remote deploy via ssh/rsync
+
+	EXAMPLES:
+
+	Install on the same host:
+
+	    $ sudo $SCRIPT_FILENAME
+
+	Deploy kernel to a remote system:
+
+	    $ sudo $SCRIPT_FILENAME -r 192.168.0.1
+
+	Deploy kernel to a VM running on this host with port 4022 forwarded and with a
+	different user name (requirement: remote user must have sudo power):
+
+	    $ sudo $SCRIPT_FILENAME -p 4022 -r localhost -u test-user
+
+	Deploy only one module to a remote system:
+
+	    $ sudo $SCRIPT_FILENAME -r remote-system -m -i 'xe.ko*'
+
+	Deploy 2 modules to a remote system:
+
+	    $ sudo $SCRIPT_FILENAME -r remote-system -m -i 'xe.ko*' -i 'i915.ko*'
+EOF
+
+}
+
+parse_arg_val() {
+	local -n ret=$1
+	shift
+	local key val="" n=0
+
+	IFS="=" read -r key val <<< "$1"
+	if [[ -z "$val" && $# -gt 1 ]]; then
+		val="$2"
+		n=$((n+1))
+	fi
+
+	# shellcheck disable=SC2034
+	ret=("$n" "$key" "$val")
+}
+
+OPT_MODULE_INCLUDE=()
+OPT_MODULE_ONLY=no
+OPT_REMOTE_USER=$(whoami)
+OPT_REMOTE_PORT=22
+OPT_REMOTE_HOST=
+parse_args() {
+	local args
+
+	while [[ $# -gt 0 ]]; do
+		case  "$1" in
+		-h|--help)
+			usage
+			exit 0
+			;;
+		-i|--include)
+			parse_arg_val args "$1" "$2"
+			shift "${args[0]}"
+			OPT_MODULE_INCLUDE+=("${args[2]}")
+			;;
+		-m|--module-only)
+			OPT_MODULE_ONLY=yes
+			;;
+		-r|--remote-host)
+			parse_arg_val args "$1" "$2"
+			shift "${args[0]}"
+			OPT_REMOTE_HOST="${args[2]}"
+			;;
+		-u|--remote-user)
+			parse_arg_val args "$1" "$2"
+			shift "${args[0]}"
+			OPT_REMOTE_USER="${args[2]}"
+			;;
+		-p|--remote-port)
+			parse_arg_val args "$1" "$2"
+			shift "${args[0]}"
+			OPT_REMOTE_PORT="${args[2]}"
+			;;
+		*)
+			argerr "Unknown option $1"
+			;;
+		esac
+		shift
+	done
+	# TODO: check arg compat/defaults
+}
+
+parse_args "$@"
+
+KREL=$(make kernelrelease)
+
+if [[ -n "$OPT_REMOTE_HOST" && "$DESTDIR" == "/" ]]; then
+	DESTDIR=$(mktemp -d tmp.install-kernel.XXXX)
+	# shellcheck disable=SC2064
+	trap "rm -rf $DESTDIR" EXIT
+fi
+
+mkdir -p "$DESTDIR/boot/"
+# TODO: figure out arch
+cp arch/x86/boot/bzImage "$DESTDIR/boot/vmlinuz-$KREL"
+make "-j$(nproc)" INSTALL_MOD_PATH="$DESTDIR" modules_install
+
+if [[ -n "$OPT_REMOTE_HOST" ]]; then
+	include_files=
+	if [[ ${#OPT_MODULE_INCLUDE[@]} -gt 0 ]]; then
+		include_files=$(printf -- "--include=%s " "${OPT_MODULE_INCLUDE[@]}")
+	fi
+
+	# copy everything to remote host
+	#
+	# shellcheck disable=SC2086
+	rsync -tcKav \
+		--chown=root:root \
+		-e "ssh -p $OPT_REMOTE_PORT" \
+		--rsync-path="sudo -u root rsync" \
+		$include_files \
+		--progress "$DESTDIR/." "$OPT_REMOTE_USER@$OPT_REMOTE_HOST:/"
+
+	if [[ "$OPT_MODULE_ONLY" == "no" ]]; then
+		typeset -f initrd -f bootloader -f sudo-function | \
+			ssh -p "$OPT_REMOTE_PORT" "$OPT_REMOTE_USER@$OPT_REMOTE_HOST" \
+			"$(cat); sudo-function initrd $KREL && sudo-function bootloader"
```

Comment:
Hm... I think this works, but wouldn't it be simpler to:

```bash
ssh -p "$OPT_REMOTE_PORT" "$OPT_REMOTE_USER@$OPT_REMOTE_HOST" \
    "sudo bash -c 'source /dev/stdin; initrd $KREL && bootloader'" < "${BASH_SOURCE[0]}"
```

With that:

- we can get rid of the sudo-function;
- we don't need to worry about maintaining the list of functions that need to be exported.

Of course, for that to work, we would also need to ensure that the "main execution" is encapsulated into a main function and gets executed only on the local machine.

Reply:
Also, because we are running sudo on the remote, wouldn't the option `-t` be required for the ssh command?

Reply:
I don't see any advantage on sending the entire script... particularly because it would execute things that it shouldn't, unless we are careful to encapsulate everything. Also the functions should still be self-contained, not relying on global variables. 

I like sending only the functions needed  -  and the reason to annotate them as "can be remotely executed" is the need to be self-contained.

Regarding `-t`, dunno... it worked for me without.

Reply:
> I don't see any advantage on sending the entire script... particularly because it would execute things that it shouldn't, unless we are careful to encapsulate everything. Also the functions should still be self-contained, not relying on global variables.
> 
> I like sending only the functions needed - and the reason to annotate them as "can be remotely executed" is the need to be self-contained.

Okay. I just don't much the fact that this solution requires us to remember to make those functions self-contained or to also include any dependencies that might arise in the future. But that works.

Reply:
> > Regarding -t, dunno... it worked for me without.

> I remember needing a pseudoterminal in some machines. Maybe this is somehow related to how sudo is configured.

Humn... that's because currently I use the script with the remote user in the sudoers without asking the password. This is not very good, but I'm not sure yet how to make it simpler without asking the passwd multiple times and having to copy files to a temporary location in the remote.

Reply:
I will leave this for future